src/modules/job-manager/plugins/dependency-after.c

```diff
@@ -0,0 +1,563 @@
+ *   jobs.
+ */
+static void release_dependency_references (flux_plugin_t *p)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    zlistx_t *l;
+    if ((l = after_refs_check (p))) {
+        struct after_ref *ref = zlistx_first (l);
+        while (ref) {
+            /*  For each after_ref entry, check to ensure the job and its
+             *   after dependencies list still exists. If so, remove this
+             *   job's entry from the list.
+             */
+            zlistx_t *after_list = after_list_check (p, ref->id);
+            if (after_list == ref->list) {
+                void *handle = zlistx_find (after_list, ref->info);
+                if (handle && zlistx_delete (ref->list, handle) < 0) {
+                    flux_log_error (h, "%s: %s: zlistx_delete",
+                                    "dependency-after",
+                                    "release_references");
+                }
+            }
+            ref = zlistx_next (l);
+        }
+    }
+
+    /*  Destroy this job's dependency reference list.
+     */
+    if (flux_jobtap_job_aux_delete (p, FLUX_JOBTAP_CURRENT_JOB, l) < 0)
+        flux_log_error (h, "release_references: flux_jobtap_job_aux_delete");
+}
+
+static int release_dependent_jobs (flux_plugin_t *p, zlistx_t *l)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    flux_job_result_t result;
+
+    if (l == NULL)
+        return 0;
+
+    if (flux_jobtap_get_job_result (p,
+                                    FLUX_JOBTAP_CURRENT_JOB,
+                                    &result) < 0) {
+        flux_log_error (h, "dependency-after: flux_jobtap_get_result");
+        return -1;
+    }
+
+    /*  Release dependent jobs based on requisite job result.
+     *   Entries will be removed from the list as they are processed.
+     */
+    if (result != FLUX_JOB_RESULT_COMPLETED)
+        release_all (p, l, AFTER_FINISH | AFTER_FAILURE);
+    else
+        release_all (p, l, AFTER_FINISH | AFTER_SUCCESS);
+
+    /*  Any remaining dependencies can't now be satisfied.
+     *   Raise exceptions on any remaining members of list `l`
+     */
+    raise_exceptions (p, l);
+
+    return 0;
+}
+
+/*  In job.state.priority, delete any dependency references in a job
+ *   that was dependent on other jobs. This prevents the references
+ *   from being released at job completion. See release_references()
+ *   call in inactive_cb() above.
+ */
+static int priority_cb (flux_plugin_t *p,
+                       const char *topic,
+                       flux_plugin_arg_t *args,
+                       void *data)
+{
+    zlistx_t *l = after_refs_check (p);
+    if (l) {
+        /*  Job has proceeded out of DEPEND state.
+         *  Delete any dependency reference list since it will no longer
+         *  be used, and we don't need to attempt deref of dependencies
+         *  (see bottom of inactive_cb()).
+         */
+        if (flux_jobtap_job_aux_delete (p, FLUX_JOBTAP_CURRENT_JOB, l) < 0) {
```

Comment:
Would it be slightly more clear here to delete using the key `flux::after_refs`? (I guess that would be aux_set with value of NULL).

I just got slightly confused wondering why this way, when the value was stored under a key.  Definitely not a big deal.

Reply:
Hm, I guess `delete` is clearer to me than `aux_set(NULL)`, however I can make this change.

Reply:
Yeah, either way is fine by me, actually.  I probably shouldn't have commented.

Reply:
the `aux_delete` call also has less useless parameters than `aux_set` with NULL. Maybe I could add a convenience macro in this case if deletion by key is more efficient, e.g.:
```
#define flux_jobtap_job_aux_destroy(p, id, key) flux_jobtap_job_aux_set (p, id, key, NULL, NULL);
```

Reply:
I think it's no more efficient - I just thought if we're naming things, it's clearer to use the names, but now I think it was fine as is and we should leave it.